common/types/list_test.go

```diff
@@ -92,58 +266,100 @@ func TestConcatListValue_Iterator(t *testing.T) {
 	}
 }
 
-func TestConcatListValue_Equal(t *testing.T) {
-	listA := NewDynamicList([]float32{1.0, 2.0}).(traits.Lister)
-	listB := NewDynamicList([]float64{3.0})
-	listConcat := listA.Add(listB)
-	// Note the internal type of list raw and concat list are slightly different.
-	listRaw := NewDynamicList([]interface{}{
-		float32(1.0), float64(2.0), float64(3.0)})
-	if listRaw.Equal(listConcat) != True ||
-		listConcat.Equal(listRaw) != True {
-		t.Errorf("Concat list and raw list were not equal, got '%v', expected '%v'",
-			listConcat.Value(),
-			listRaw.Value())
+func TestStringList_Add_Empty(t *testing.T) {
+	list := NewStringList([]string{"hello"})
+	if list.Add(NewStringList([]string{})) != list {
+		t.Error("Adding empty lists resulted in new list creation.")
 	}
-	if listRaw.Equal(listA) == True {
-		t.Errorf("Lists of unequal length considered equal")
+	if NewStringList([]string{}).Add(list) != list {
+		t.Error("Adding empty lists resulted in new list creation.")
 	}
 }
 
-func TestListValue_Contains(t *testing.T) {
-	listA := NewDynamicList([]float32{1.0, 2.0}).(traits.Lister)
-	listB := NewDynamicList([]float64{3.0})
-	list := listA.Add(listB).(traits.Lister).Add(listA).(traits.Lister)
-	if list.Contains(Int(3)) == True {
-		t.Error("List contains succeeded with wrong type")
+func TestStringList_Add_Error(t *testing.T) {
+
+}
+
+func TestStringList_Add_Heterogenous(t *testing.T) {
+	listA := NewStringList([]string{"hello"})
+	listB := NewDynamicList([]int32{1, 2, 3})
+	list := listA.Add(listB).(traits.Lister).Value().([]interface{})
+	expected := []interface{}{"hello", int32(1), int32(2), int32(3)}
+	if len(list) != len(expected) {
+		t.Error("Unexpected list size. Got '%d', expected 4", len(list))
 	}
-	if list.Contains(Double(3)) != True {
-		t.Error("List contains did not succeed")
+	for i, v := range expected {
+		if list[i] != v {
+			t.Errorf("elem[%d] Got '%v', expected '%v'", i, list[i], v)
+		}
 	}
 }
 
-func TestListValue_ToProto(t *testing.T) {
-	listA := NewDynamicList([]float32{1.0, 2.0}).(traits.Lister)
-	listB := NewDynamicList([]float64{3.0})
-	listConcat := listA.Add(listB)
-	if protoList, err := listConcat.ConvertToNative(reflect.TypeOf([]float32{})); err != nil {
+func TestStringList_Add_StringLists(t *testing.T) {
+	listA := NewStringList([]string{"hello"})
+	listB := NewStringList([]string{"world", "!"})
+	list := listA.Add(listB).(traits.Lister)
+	if list.Size() != Int(3) {
+		t.Error("Combined list did not have correct size.")
+	}
+	expected := []string{"hello", "world", "!"}
+	for i, v := range expected {
+		if list.Get(Int(i)).Equal(String(v)) != True {
+			t.Error("elem[%d] Got '%v', expected '%v'", i, list.Get(Int(i)), v)
+		}
+	}
+}
+
+func TestStringList_ConvertToNative(t *testing.T) {
+	list := NewStringList([]string{"h", "e", "l", "p"})
+	_, err := list.ConvertToNative(reflect.TypeOf([]string{}))
+	if err != nil {
+		t.Error("Unable to convert string list to itself.")
+	}
```

Comment:
Do we need to check if the converted content is correct?

Reply:
Done.